File: Samples/RadialMenu/UE4/Source/RadialMenu/RadialItem.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "RadialItem.h"

#include <NsCore/ReflectionImplement.h>
#include <NsGui/UIElementData.h>
#include <NsGui/FrameworkPropertyMetadata.h>


using namespace RadialMenu;


////////////////////////////////////////////////////////////////////////////////////////////////////
RadialItem::RadialItem()
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
RadialItem::~RadialItem()
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialItem::GetCenterX() const
{
    return GetValue<float>(CenterXProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialItem::SetCenterX(float value)
{
    SetValue<float>(CenterXProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialItem::GetCenterY() const
{
    return GetValue<float>(CenterYProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialItem::SetCenterY(float value)
{
    SetValue<float>(CenterYProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialItem::GetRadius() const
{
    return GetValue<float>(RadiusProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialItem::SetRadius(float value)
{
    SetValue<float>(RadiusProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialItem::GetOffset() const
{
    return GetValue<float>(OffsetProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialItem::SetOffset(float value)
{
    SetValue<float>(OffsetProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialItem::GetAngleStart() const
{
    return GetValue<float>(AngleStartProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialItem::SetAngleStart(float value)
{
    SetValue<float>(AngleStartProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialItem::GetAngleDelta() const
{
    return GetValue<float>(AngleDeltaProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialItem::SetAngleDelta(float value)
{
    SetValue<float>(AngleDeltaProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
static Noesis::Point RadialPoint(const Noesis::Point& center, float angle, float radius)
{
    float radiansAngle = (angle - 90.0f) * Noesis::DegToRad;
    float x = radius * cosf(radiansAngle);
    float y = radius * sinf(radiansAngle);
    return Noesis::Point(center.x + x, center.y + y);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::Size RadialItem::ArrangeOverride(const Noesis::Size& finalSize)
{
    Noesis::UIElement* child = GetChild();
    if (child != nullptr)
    {
        float angleStart = GetAngleStart();
        float angleDelta = GetAngleDelta();
        float radius = GetRadius();
        Noesis::Point center(GetCenterX(), GetCenterY());
        Noesis::Point point = RadialPoint(center, angleStart + angleDelta * 0.5f, radius);

        const Noesis::Size& childSize = child->GetDesiredSize();
        float halfChildWidth = childSize.width / 2.0f;
        float halfChildHeight = childSize.height / 2.0f;

        Noesis::Point p0(point.x - halfChildWidth, point.y - halfChildHeight);
        Noesis::Point p1(point.x + halfChildWidth, point.y + halfChildHeight);
        Noesis::Rect childRect;
        childRect.SetLocation(p0);
        childRect.Expand(p1);

        child->Arrange(childRect);
    }

    return finalSize;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(RadialItem, "RadialMenu.RadialItem")
{
    Noesis::UIElementData* data = NsMeta<Noesis::UIElementData>(Noesis::TypeOf<SelfClass>());

    data->RegisterProperty<float>(CenterXProperty, "CenterX",
        Noesis::FrameworkPropertyMetadata::Create(0.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
    data->RegisterProperty<float>(CenterYProperty, "CenterY",
        Noesis::FrameworkPropertyMetadata::Create(0.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
    data->RegisterProperty<float>(RadiusProperty, "Radius",
        Noesis::FrameworkPropertyMetadata::Create(0.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
    data->RegisterProperty<float>(OffsetProperty, "Offset",
        Noesis::FrameworkPropertyMetadata::Create(0.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
    data->RegisterProperty<float>(AngleStartProperty, "AngleStart",
        Noesis::FrameworkPropertyMetadata::Create(0.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
    data->RegisterProperty<float>(AngleDeltaProperty, "AngleDelta",
        Noesis::FrameworkPropertyMetadata::Create(0.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Noesis::DependencyProperty* RadialItem::CenterXProperty;
const Noesis::DependencyProperty* RadialItem::CenterYProperty;
const Noesis::DependencyProperty* RadialItem::RadiusProperty;
const Noesis::DependencyProperty* RadialItem::OffsetProperty;
const Noesis::DependencyProperty* RadialItem::AngleStartProperty;
const Noesis::DependencyProperty* RadialItem::AngleDeltaProperty;

File: Samples/RadialMenu/UE4/Assets/RadialMenu/RadialMenu/RadialItem.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Shapes;

namespace RadialMenu
{
    /// <summary>
    /// Panel to display items in a radial/arc.
    /// </summary>
    public class RadialItem : Decorator
    {
        #region Dependency Properties
        /// <summary>
        /// The center X coordinate of the circle from which this element is positioned
        /// </summary>
        public double CenterX
        {
            get { return (double)GetValue(CenterXProperty); }
            set { SetValue(CenterXProperty, value); }
        }

        public static readonly DependencyProperty CenterXProperty = DependencyProperty.Register(
            "CenterX", typeof(double), typeof(RadialItem), new FrameworkPropertyMetadata(0.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        /// <summary>
        /// The center Y coordinate of the circle from which this element is positioned
        /// </summary>
        public double CenterY
        {
            get { return (double)GetValue(CenterYProperty); }
            set { SetValue(CenterYProperty, value); }
        }

        public static readonly DependencyProperty CenterYProperty = DependencyProperty.Register(
            "CenterY", typeof(double), typeof(RadialItem), new FrameworkPropertyMetadata(0.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        /// <summary>
        /// The radius where this element is positioned
        /// </summary>
        public double Radius
        {
            get { return (double)GetValue(RadiusProperty); }
            set { SetValue(RadiusProperty, value); }
        }

        public static readonly DependencyProperty RadiusProperty = DependencyProperty.Register(
            "Radius", typeof(double), typeof(RadialItem), new FrameworkPropertyMetadata(0.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        /// <summary>
        /// The distance to translate this element out from the center
        /// </summary>
        public double Offset
        {
            get { return (double)GetValue(OffsetProperty); }
            set { SetValue(OffsetProperty, value); }
        }

        public static readonly DependencyProperty OffsetProperty = DependencyProperty.Register(
            "Offset", typeof(double), typeof(RadialItem), new FrameworkPropertyMetadata(0.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        /// <summary>
        /// The start angle from the Y axis vector of this element in degrees
        /// </summary>
        public double AngleStart
        {
            get { return (double)GetValue(AngleStartProperty); }
            set { SetValue(AngleStartProperty, value); }
        }

        public static readonly DependencyProperty AngleStartProperty = DependencyProperty.Register(
            "AngleStart", typeof(double), typeof(RadialItem), new FrameworkPropertyMetadata(0.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        /// <summary>
        /// The angle delta from the Y axis vector of this element in degrees
        /// </summary>
        public double AngleDelta
        {
            get { return (double)GetValue(AngleDeltaProperty); }
            set { SetValue(AngleDeltaProperty, value); }
        }

        public static readonly DependencyProperty AngleDeltaProperty = DependencyProperty.Register(
            "AngleDelta", typeof(double), typeof(RadialItem), new FrameworkPropertyMetadata(0.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        #endregion

        static RadialItem()
        {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(RadialItem), new FrameworkPropertyMetadata(typeof(RadialItem)));
        }

        protected override Size ArrangeOverride(Size arrangeSize)
        {
            UIElement child = Child;
            if (child != null)
            {
                Point center = new Point(CenterX, CenterY);
                Point point = RadialPoint(center, AngleStart + AngleDelta * 0.5, Radius);

                var halfChildWidth = child.DesiredSize.Width / 2;
                var halfChildHeight = child.DesiredSize.Height / 2;

                var p0 = new Point(point.X - halfChildWidth, point.Y - halfChildHeight);
                var p1 = new Point(point.X + halfChildWidth, point.Y + halfChildHeight);
                var childRect = new Rect(p0, p1);

                child.Arrange(childRect);
            }

            return arrangeSize;
        }

        private static Point RadialPoint(Point center, double angle, double radius)
        {
            // Converts to radians
            double radiansAngle = (angle - 90) * Math.PI / 180.0;
            double x = radius * Math.Cos(radiansAngle);
            double y = radius * Math.Sin(radiansAngle);
            return new Point(center.X + x, center.Y + y);
        }
    }
}
